src/porepy/numerics/nonlinear/convergence_check.py

```diff
@@ -0,0 +1,455 @@
+        """Reset the reference values to their initial state."""
+        self.reference_value.reset()
+
+    ### Convergence check methods
+
+    def _check(
+        self,
+        nonlinear_increment_norm: dict[str, float],
+        residual_norm: dict[str, float],
+        tol: ConvergenceTolerance,
+    ) -> tuple[ConvergenceStatus, ConvergenceInfo]:
+        """Check convergence using relative norms.
+
+        Parameters:
+            nonlinear_increment_norm: Norm of the nonlinear increment.
+            residual_norm: Norm of the residual.
+            ConvergenceTolerance: Tolerances for the convergence check.
+
+        Returns:
+            ConvergenceStatus: Convergence status of the non-linear iteration.
+            ConvergenceInfo: Information about the convergence check.
+
+        """
+        # Consistency checks
+        self._check_dicts(
+            nonlinear_increment_norm, self.reference_nonlinear_increment_norm
+        )
+        self._check_dicts(residual_norm, self.reference_residual_norm)
+
+        # Check nan.
+        is_nan = any(np.isnan(res_norm) for res_norm in residual_norm.values()) or any(
+            np.isnan(inc_norm) for inc_norm in nonlinear_increment_norm.values()
+        )
+
+        # Check divergence.
+        is_diverged = any(
+            res_norm > tol.max_residual for res_norm in residual_norm.values()
+        ) or any(
+            inc_norm > tol.max_increment
+            for inc_norm in nonlinear_increment_norm.values()
+        )
+
+        # Reduce norms to floats using l-infinity norm over combined
+        # absolute-relative values.
+        reduced_nonlinear_increment_norm = max(
+            inc_norm / (1 + self.reference_nonlinear_increment_norm[key])
+            for key, inc_norm in nonlinear_increment_norm.items()
+        )
+        reduced_residual_norm = max(
+            res_norm / (1 + self.reference_residual_norm[key])
+            for key, res_norm in residual_norm.items()
+        )
+
+        # Check convergence using relative norms.
+        converged_inc = reduced_nonlinear_increment_norm < tol.tol_increment
+        converged_res = reduced_residual_norm < tol.tol_residual
+        is_converged = converged_inc and converged_res
+
+        # Determine convergence status.
+        convergence_status = ConvergenceStatus.NOT_CONVERGED
+        if is_nan:
+            convergence_status = ConvergenceStatus.NAN
+        elif is_diverged:
+            convergence_status = ConvergenceStatus.DIVERGED
+        elif is_converged:
+            convergence_status = ConvergenceStatus.CONVERGED
+
+        # Collect information about the convergence check.
+        info = ConvergenceInfo(
+            nonlinear_increment_norm=reduced_nonlinear_increment_norm,
+            residual_norm=reduced_residual_norm,
+        )
+
+        return convergence_status, info
+
+
+### Concrete convergence criteria
+
+
+# class NanConvergenceCriterion(ConvergenceCriterion):
```

Comment:
I have a feeling we want to check for NANs (almost) always. Does that indicate a need to be able to combine Criteria?